File: python/213. House Robber II.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        res = [0] * (len(nums) - 1)
        res[0] = nums[0]
        res[1] = max(nums[0], nums[1])
        for i in range(2, len(nums)-1):
            res[i] = max(nums[i] + res[i-2], res[i-1])
        res1 = res[-1]
        res = [0] * (len(nums) - 1)
        res[0] = nums[1]
        res[1] = max(nums[2], nums[1])
        for i in range(2, len(nums) - 1):
            res[i] = max(nums[i+1] + res[i-2], res[i-1])
        res2 = res[-1]
        return max(res1, res2)

File: cpp/201-250/213. House Robber II.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty())
            return 0;
        if (nums.size() == 1)
            return nums[0];
        int prev = 0, cur = 0, res = 0;
        for (int i = 0; i < nums.size() - 1; i++) {
            auto tmp = cur;
            cur = max(cur, prev + nums[i]);
            prev = tmp;
        }
        res = cur;
        prev = 0, cur = 0;
        for(int i = 1; i < nums.size(); i++){
            auto tmp = cur;
            cur = max(cur, prev + nums[i]);
            prev = tmp;
        }
        return max(res, cur);
    }
};